gato/training/trainer.py

```diff
@@ -116,13 +123,31 @@ def train_step(self):
 
             if not self.args.disable_grad_clip and self.accelerator.sync_gradients:
                 self.accelerator.clip_grad_norm_(self.model.parameters(), self.args.grad_norm_clip)
-            
+
             self.optimizer.step()
             self.scheduler.step()
             self.optimizer.zero_grad()
 
         return loss.detach().cpu().item(), logs
 
+    def process_text_batch(self, batch_size):
+        # Select batch_size number of random indices from the size of the text dataset.
+        random_indices = np.random.randint(0, len(self.text_dataset['train']), size=batch_size)
+        selected_text_examples = [self.text_dataset['train'][idx]['text'] for idx in random_indices]
+
+        input_dict = {}
+
+        text_tokens = [self.model.text_tokenizer.encode(text) for text in selected_text_examples]
+        input_dict['text'] = torch.stack(text_tokens).to(self.args.device)
+        # The targets are the same as the input, but shifted one position to the left
+        # append the appropriate end-of-sequence token as well
+        input_dict['text_targets'] = torch.cat(
+            [tokens[1:], torch.tensor([self.model.text_tokenizer.tokenizer.eos_token_id]).expand(1, tokens.shape[1])],
+            dim=0
+        ).to(self.args.device)
+
+        return input_dict
```

Comment:
Text is already tokenized (text_tokenizer.encode), right now, as implemented in tokenize_input_dicts(), it would try to tokenize a second time.